scripts/lib/CIME/XML/component.py

```diff
@@ -117,12 +122,96 @@ def _get_value_match(self, node, attributes=None, exact_match=False):
 
     #pylint: disable=arguments-differ
     def get_description(self, compsetname):
+        if self.get_version() == 3.0:
+            return self._get_description_v3(compsetname, comp_class=self._comp_class)
+        else:
+            return self._get_description_v2(compsetname)
+
+    def get_forcing_description(self, compsetname):
+        if self.get_version() == 3.0:
+            return self._get_description_v3(compsetname, comp_class='forcing')
+        else:
+            return ""
+
+
+    def _get_description_v3(self, compsetname, comp_class=None):
+        comp_class = comp_class.lower()
         rootnode = self.get_node("description")
+        desc = ""
         desc_nodes = self.get_nodes("desc", root=rootnode)
+
+        expect(comp_class is not None,"comp_class argument required for version3 files")
+        modifier_mode = rootnode.get('modifier_mode')
+        if modifier_mode is None:
+            modifier_mode = '*'
+        expect(modifier_mode in ('*','1','?','+'),
+               "Invalid modifier_mode {} in file {}".format(modifier_mode, self.filename))
+
+        optiondesc = {}
+        if comp_class == "forcing":
+            for node in desc_nodes:
+                forcing = node.get('forcing')
+                if forcing is not None and compsetname.startswith(forcing+'_'):
+                    expect(len(desc)==0,
+                           "Too many matches on forcing field {} in file {}".\
+                               format(forcing, self.filename))
+                    desc = node.text
+
+            expect(len(desc) > 0,"No match found for forcing field {} in file {}".\
+                       format(forcing, self.filename))
+            return desc
+
+        # first pass just make a hash of the option descriptions
+        for node in desc_nodes:
+            option = node.get('option')
+            if option is not None:
+                optiondesc[option] = node.text
+        #second pass find a comp_class match
```

Comment:
This is the one comment from before that you didn't resolve:

> I couldn't understand what the following code is doing, at a high level (so I haven't given it much of a review). This could use some high-level documentation and/or examples. Ideally, there would also be some unit tests - either for this whole get_description method or for the core algorithm (below), which could be extracted into its own method.